src/sched/lthread.c

```diff
@@ -474,23 +476,32 @@ int __init_utp(void *p, int set_tp)
 void *__copy_utls(struct lthread *lt, unsigned char *mem, size_t sz)
 {
 	mem += sz - sizeof(struct lthread_tcb_base);
-	mem -= (uintptr_t)mem & (libc.tls_align - 1);
+	mem -= (uintptr_t)mem & (TLS_ALIGN - 1);
 	return (void *)mem;
 }
 
 void set_tls_tp(struct lthread* lt)
 {
+    uintptr_t tp_unaligned, tp;
     if (!lt->itls)
         return;
-
-    uintptr_t tp_unaligned =
+    // pthread_create passes thread pointer for lt->itls
+    // no need to calculate offset of tp in tls region
+    if (lt->attr.stack_size == 0)
+    {
+        tp = lt->itls;
+    }
+    else
+    {
+        tp_unaligned =
         (uintptr_t)(lt->itls + lt->itlssz - sizeof(struct lthread_tcb_base));
-    struct lthread_tcb_base* tp =
-        (struct
-         lthread_tcb_base*)(tp_unaligned - (tp_unaligned & (libc.tls_align - 1)));
-
-    tp->schedctx = __scheduler_self();
+        tp =
+            (struct
+            lthread_tcb_base*)(tp_unaligned - (tp_unaligned & (TLS_ALIGN - 1)));
 
+        tp->schedctx = __scheduler_self();
```

Comment:
This should no longer be required.  I don't think we ever read this field, we should always be getting the scheduler via %gs.  If you remove this, does anything break?

Reply:
Should be safe to remove, will do so.